### PR TITLE
Add documentation

### PR DESCRIPTION
**Pull request by victorhmp**
#### What problem is this solving?

Currently, we don't have any documentation on this app, even though it is heavily used. This should make it a bit easier to use. 

#### How to test it?

You can see the rendered docs at https://github.com/vtex-apps/product-context/tree/feature/documentation.

#### Related to / Depends on

Related to: https://github.com/vtex-apps/product-context/pull/32.

### CODE REVIEW

File: docs/README.md [hunk 1]
```diff
@@ -1,7 +1,132 @@
...
+
+  return (
+    <Fragment>
+      {productContextValue?.product?.productName}
+    </Fragment>
+  )
+}
+
+export default MyComponent
+```
```
Comment (klzns) [inline docs/README.md:40]: If we are going to document the usage in TypeScript I think we should make it clear that this file is a TypeScript file. People will copy and paste this code and will not work.

File: docs/README.md [hunk 2]
```diff
@@ -1,7 +1,132 @@
...
+
+:warning: Be sure to run `vtex setup` in your project to install the correct TypeScript types exported by this app.
+
+## API
+
+### `ProductContext`
+
+The `ProductContext` entry point exports the React context in it's raw form. This is only useful if you want to reference it directly.
+
+Be aware that most of the times, what you're looking for is the `useProduct` hook described below.
```
Comment (klzns) [inline docs/README.md:50]: Move this bellow of `useProduct`

File: docs/README.md [hunk 3]
```diff
@@ -1,7 +1,132 @@
...
+It expects to receive the following props:
+
+```ts
+interface ProductAndQuery {
+  query: Record<string, any>
+  product: MaybeProduct
+}
+```
+
+and will also wrap it's children with the `ProductDispatchContext` component described below.
```
Comment (klzns) [inline docs/README.md:92]: Not sure if we should document this. I can't see a clear use case.

File: docs/README.md [hunk 4]
```diff
@@ -1,7 +1,132 @@
...
+
+### `ProductDispatchContext`
+
+The `ProductDispatchContext` entry point exports a React context in it's raw form. This is only useful if you want to reference it directly. Most likely you don't need to use this, since the `ProductContextProvider` already wraps it's children with a provider for this context.
+
+This is the context that exposes functions you can use to manipulate the nearest `ProductContext`. It's value is a `dispatch` function, capable of performing the following `actions`:
```
Comment (klzns) [inline docs/README.md:98]: I would not expose this and leave just a hook like `useProductDispatch()`

File: docs/README.md (hunk 3 above)
Comment (victorhmp) [reply docs/README.md:92]: Yep, agreed

File: docs/README.md [hunk 5]
```diff
@@ -23,6 +23,7 @@ To use this app, be sure to add it to you app's `manifest.json` file:
 then you can import any of the exported components and hooks from the app. Here's an example of a component that render's the name of the product whose data is stored in the nearest `ProductContext`:
 
 ```tsx
+// Notice that this is TypeScript, and this code should be in a .tsx file
```
Comment (klzns) [inline docs/README.md:26]: :notbad:

File: docs/README.md [hunk 6]
```diff
@@ -76,26 +71,15 @@ you should expect an object that looks like that as the return value of `useProd
...
-
-It expects to receive the following props:
-
-```ts
-interface ProductAndQuery {
-  query: Record<string, any>
-  product: MaybeProduct
-}
-```
+### `ProductContext`
```
Comment (klzns) [inline docs/README.md:74]: I think we can remove this, what do you think?

File: docs/README.md [hunk 7]
```diff
@@ -76,26 +71,15 @@ you should expect an object that looks like that as the return value of `useProd
...
+Be aware that most of the times, what you're looking for is the `useProduct` hook described below.
 
-The `ProductDispatchContext` entry point exports a React context in it's raw form. This is only useful if you want to reference it directly. Most likely you don't need to use this, since the `ProductContextProvider` already wraps it's children with a provider for this context.
+### `useProductDispatch`
 
-This is the context that exposes functions you can use to manipulate the nearest `ProductContext`. It's value is a `dispatch` function, capable of performing the following `actions`:
+This hooks returns a `dispatch` function you can use to manipulate the nearest `ProductContext`. This function is capable of performing the following `actions`:
 
 ```ts
 type Actions =
```
Comment (klzns) [inline docs/README.md:85]: This is not legible

File: docs/README.md (hunk 6 above)
Comment (kaisermann) [reply docs/README.md:74]: Agreed, if people start needing it we can make it public.

File: docs/README.md (hunk 7 above)
Comment (iaronaraujo) [reply docs/README.md:85]: What do you mean @klzns ?

Comment (kaisermann) [reply docs/README.md:85]: I tried to move each one to its own type, but it got way worse: 

```ts
type ACTION_SELECT_IMAGE_VARIATION = Action<
  'SELECT_IMAGE_VARIATION',
  { args: { selectedImageVariationSKU: string | null } }
>;
type ACTION_SET_QUANTITY = Action<
  'SET_QUANTITY',
  { args: { quantity: number } }
>;
type ACTION_SKU_SELECTOR_SET_VARIATIONS_SELECTED = Action<
  'SKU_SELECTOR_SET_VARIATIONS_SELECTED',
  { args: { allSelected: boolean } }
>;
type ACTION_SET_BUY_BUTTON_CLICKED = Action<
  'SET_BUY_BUTTON_CLICKED',
  { args: { clicked: boolean } }
>;
type ACTION_SKU_SELECTOR_SET_IS_VISIBLE = Action<
  'SKU_SELECTOR_SET_IS_VISIBLE',
  { args: { isVisible: boolean } }
>;
type ACTION_SET_SELECTED_ITEM = Action<
  'SET_SELECTED_ITEM',
  { args: { item: Item | undefined | null } }
>;
type ACTION_SET_ASSEMBLY_OPTIONS = Action<
  'SET_ASSEMBLY_OPTIONS',
  {
    args: {
      groupId: string;
      groupItems: AssemblyOptionItem[];
      groupInputValues: InputValues;
      isValid: boolean;
    };
  }
>;
type ACTION_SET_PRODUCT = Action<
  'SET_PRODUCT',
  { args: { product: MaybeProduct } }
>;

type Actions =
  | ACTION_SELECT_IMAGE_VARIATION
  | ACTION_SET_QUANTITY
  | ACTION_SKU_SELECTOR_SET_VARIATIONS_SELECTED
  | ACTION_SET_BUY_BUTTON_CLICKED
  | ACTION_SKU_SELECTOR_SET_IS_VISIBLE
  | ACTION_SET_SELECTED_ITEM
  | ACTION_SET_ASSEMBLY_OPTIONS
  | ACTION_SET_PRODUCT;
```

Comment (kaisermann) [reply docs/README.md:85]: Maybe you could just specify the action names?

Comment (klzns) [reply docs/README.md:85]: I suggest to document as markdown text

File: docs/README.md (hunk 6 above)
Comment (victorhmp) [reply docs/README.md:74]: Yep, agreed

**Comment by victorhmp** [comment]
@klzns @kaisermann I updated the `useProductDispatch` section to make it simpler. How do you guys like it?

**Review by Klynger — APPROVED** [review]

**Pull request merged**